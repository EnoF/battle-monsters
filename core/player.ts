export interface Player {
  hp: number;
  maxHp: number;
  move?: PlayerMove;
  moves: Array<PlayerMove>;
}
export enum PlayerMoveType {
  ATTACK = "ATTACK",
  PARRY = "PARRY",
  DODGE = "DODGE",
  STAGGER = "STAGGER",
}
export interface PlayerMove {
  type: PlayerMoveType;
  power?: number;
  aoe?: number;
  shift?: number;
  turns?: number;
  cooldown?: number;
}
export const createCharacter = (): Player => ({
  hp: 10,
  maxHp: 10,
  moves: [
    {
      type: PlayerMoveType.ATTACK,
      power: 1,
    },
    {
      type: PlayerMoveType.ATTACK,
      power: 1,
      shift: 1,
    },
    {
      type: PlayerMoveType.ATTACK,
      power: 3,
    },
    {
      type: PlayerMoveType.ATTACK,
      power: 3,
      shift: 1,
    },
    {
      type: PlayerMoveType.PARRY,
    },
    {
      type: PlayerMoveType.DODGE,
    },
  ],
});

export const getAoe = (move: PlayerMove) => move.aoe || 0;
export const getShift = (move: PlayerMove) => move.shift || 0;
export const getPower = (move: PlayerMove) => move.power || 0;
export const getCooldown = (move: PlayerMove) => move.cooldown || 0;
export const isStaggered = (player: Player) =>
  player.move?.type === PlayerMoveType.STAGGER && player.move?.turns > 0;
