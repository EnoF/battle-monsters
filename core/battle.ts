import { Player, PlayerMoveType, getAoe, getShift } from "./player";

export interface BattleResult {
  p1: Player;
  p2: Player;
}
export interface BattleConfig {
  p1: Player;
  p2: Player;
}

const calculateBattle = ({ move, ...player }: Player, opponent: Player) => {
  switch (opponent.move.type) {
    case PlayerMoveType.PARRY:
      if (move.type !== PlayerMoveType.ATTACK) return player;
      const stagnation = move.power + getShift(move) + getAoe(move) - 1;
      if (stagnation <= 0) return player;
      return {
        ...player,
        move: {
          type: PlayerMoveType.STAGGER,
          turns: stagnation,
        },
      };
    case PlayerMoveType.ATTACK:
      if (move.type === PlayerMoveType.DODGE) return player;
      if (move.type === PlayerMoveType.PARRY) return player;
      if (getShift(move) > getAoe(opponent.move)) return player;
      return { ...player, hp: player.hp - opponent.move.power };
    case PlayerMoveType.DODGE:
      if (move.type !== PlayerMoveType.PARRY) return player;
      return {
        ...player,
        move: {
          type: PlayerMoveType.STAGGER,
          turns: 1,
        },
      };
    default:
      return player;
  }
};
function battle(battleConfig: BattleConfig): BattleResult {
  return {
    p1: calculateBattle(battleConfig.p1, battleConfig.p2),
    p2: calculateBattle(battleConfig.p2, battleConfig.p1),
  };
}
export default battle;
