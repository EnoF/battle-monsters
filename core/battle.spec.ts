import { getPermutations } from "./permutations";
import battle, { BattleConfig, BattleResult } from "./battle";
import { PlayerMoveType } from "./player";

const getPowerDescription = (move) => {
  if (!move.power) return "";
  return `with ${move.power} power`;
};
const testBattleRound = (
  battleConfig: BattleConfig,
  expectedResult: BattleResult
) => {
  describe(`when player 1: ${battleConfig.p1.move.type} ${getPowerDescription(
    battleConfig.p1.move
  )}`, () => {
    describe(`and player 2: ${battleConfig.p2.move.type} ${getPowerDescription(
      battleConfig.p2.move
    )}`, () => {
      if (expectedResult)
        return it.skip("should verify new test case", () => {
          expect(battle(battleConfig)).toMatchObject(expectedResult);
        });
      it("calculates it correctly", () => {
        expect(battle(battleConfig)).toMatchSnapshot();
      });
    });
  });
};

describe("Battle system", () => {
  const availableMoves = [
    { type: PlayerMoveType.ATTACK, power: 1 },
    { type: PlayerMoveType.ATTACK, power: 2 },
    { type: PlayerMoveType.ATTACK, power: 3 },
    { type: PlayerMoveType.ATTACK, power: 1, shift: 1 },
    { type: PlayerMoveType.ATTACK, power: 1, aoe: 1 },
    { type: PlayerMoveType.ATTACK, power: 3, aoe: 2, shift: 2 },
    { type: PlayerMoveType.BLOCK },
    { type: PlayerMoveType.DODGE },
  ];
  getPermutations({
    p1Hp: [10],
    p2Hp: [10],
    p1Move: availableMoves,
    p2Move: availableMoves,
  }).forEach(({ p1Hp, p2Hp, p1Move, p2Move }) =>
    testBattleRound({
      p1: { hp: p1Hp, move: p1Move },
      p2: { hp: p2Hp, move: p2Move },
    })
  );
  testBattleRound(
    {
      p1: {
        hp: 10,
        move: { type: PlayerMoveType.ATTACK, aoe: 2, power: 3, shift: 2 },
      },
      p2: {
        hp: 10,
        move: { type: PlayerMoveType.ATTACK, power: 1, shift: 1 },
      },
    },
    {
      p1: { hp: 10 },
      p2: { hp: 7 },
    }
  );
});
