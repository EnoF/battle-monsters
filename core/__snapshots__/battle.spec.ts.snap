// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Battle system when player 1: ATTACK with 1 power and player 2: ATTACK with 1 power calculates it correctly 1`] = `
Object {
  "p1": Object {
    "hp": 9,
  },
  "p2": Object {
    "hp": 9,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 1 power and player 2: ATTACK with 1 power calculates it correctly 2`] = `
Object {
  "p1": Object {
    "hp": 9,
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 1 power and player 2: ATTACK with 1 power calculates it correctly 3`] = `
Object {
  "p1": Object {
    "hp": 9,
  },
  "p2": Object {
    "hp": 9,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 1 power and player 2: ATTACK with 1 power calculates it correctly 4`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 9,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 1 power and player 2: ATTACK with 1 power calculates it correctly 5`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 1 power and player 2: ATTACK with 1 power calculates it correctly 6`] = `
Object {
  "p1": Object {
    "hp": 9,
  },
  "p2": Object {
    "hp": 9,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 1 power and player 2: ATTACK with 1 power calculates it correctly 7`] = `
Object {
  "p1": Object {
    "hp": 9,
  },
  "p2": Object {
    "hp": 9,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 1 power and player 2: ATTACK with 1 power calculates it correctly 8`] = `
Object {
  "p1": Object {
    "hp": 9,
  },
  "p2": Object {
    "hp": 9,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 1 power and player 2: ATTACK with 1 power calculates it correctly 9`] = `
Object {
  "p1": Object {
    "hp": 9,
  },
  "p2": Object {
    "hp": 9,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 1 power and player 2: ATTACK with 2 power calculates it correctly 1`] = `
Object {
  "p1": Object {
    "hp": 8,
  },
  "p2": Object {
    "hp": 9,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 1 power and player 2: ATTACK with 2 power calculates it correctly 2`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 9,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 1 power and player 2: ATTACK with 2 power calculates it correctly 3`] = `
Object {
  "p1": Object {
    "hp": 8,
  },
  "p2": Object {
    "hp": 9,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 1 power and player 2: ATTACK with 3 power calculates it correctly 1`] = `
Object {
  "p1": Object {
    "hp": 7,
  },
  "p2": Object {
    "hp": 9,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 1 power and player 2: ATTACK with 3 power calculates it correctly 2`] = `
Object {
  "p1": Object {
    "hp": 7,
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 1 power and player 2: ATTACK with 3 power calculates it correctly 3`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 9,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 1 power and player 2: ATTACK with 3 power calculates it correctly 4`] = `
Object {
  "p1": Object {
    "hp": 7,
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 1 power and player 2: ATTACK with 3 power calculates it correctly 5`] = `
Object {
  "p1": Object {
    "hp": 7,
  },
  "p2": Object {
    "hp": 9,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 1 power and player 2: ATTACK with 3 power calculates it correctly 6`] = `
Object {
  "p1": Object {
    "hp": 7,
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 1 power and player 2: DODGE  calculates it correctly 1`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 1 power and player 2: DODGE  calculates it correctly 2`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 1 power and player 2: DODGE  calculates it correctly 3`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 1 power and player 2: PARRY  calculates it correctly 1`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 1 power and player 2: PARRY  calculates it correctly 2`] = `
Object {
  "p1": Object {
    "hp": 10,
    "move": Object {
      "turns": 1,
      "type": "STAGGER",
    },
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 1 power and player 2: PARRY  calculates it correctly 3`] = `
Object {
  "p1": Object {
    "hp": 10,
    "move": Object {
      "turns": 1,
      "type": "STAGGER",
    },
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 2 power and player 2: ATTACK with 1 power calculates it correctly 1`] = `
Object {
  "p1": Object {
    "hp": 9,
  },
  "p2": Object {
    "hp": 8,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 2 power and player 2: ATTACK with 1 power calculates it correctly 2`] = `
Object {
  "p1": Object {
    "hp": 9,
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 2 power and player 2: ATTACK with 1 power calculates it correctly 3`] = `
Object {
  "p1": Object {
    "hp": 9,
  },
  "p2": Object {
    "hp": 8,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 2 power and player 2: ATTACK with 2 power calculates it correctly 1`] = `
Object {
  "p1": Object {
    "hp": 8,
  },
  "p2": Object {
    "hp": 8,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 2 power and player 2: ATTACK with 3 power calculates it correctly 1`] = `
Object {
  "p1": Object {
    "hp": 7,
  },
  "p2": Object {
    "hp": 8,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 2 power and player 2: ATTACK with 3 power calculates it correctly 2`] = `
Object {
  "p1": Object {
    "hp": 7,
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 2 power and player 2: DODGE  calculates it correctly 1`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 2 power and player 2: PARRY  calculates it correctly 1`] = `
Object {
  "p1": Object {
    "hp": 10,
    "move": Object {
      "turns": 1,
      "type": "STAGGER",
    },
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 3 power and player 2: ATTACK with 1 power calculates it correctly 1`] = `
Object {
  "p1": Object {
    "hp": 9,
  },
  "p2": Object {
    "hp": 7,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 3 power and player 2: ATTACK with 1 power calculates it correctly 2`] = `
Object {
  "p1": Object {
    "hp": 9,
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 3 power and player 2: ATTACK with 1 power calculates it correctly 3`] = `
Object {
  "p1": Object {
    "hp": 9,
  },
  "p2": Object {
    "hp": 7,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 3 power and player 2: ATTACK with 1 power calculates it correctly 4`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 7,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 3 power and player 2: ATTACK with 1 power calculates it correctly 5`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 7,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 3 power and player 2: ATTACK with 1 power calculates it correctly 6`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 7,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 3 power and player 2: ATTACK with 2 power calculates it correctly 1`] = `
Object {
  "p1": Object {
    "hp": 8,
  },
  "p2": Object {
    "hp": 7,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 3 power and player 2: ATTACK with 2 power calculates it correctly 2`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 7,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 3 power and player 2: ATTACK with 3 power calculates it correctly 1`] = `
Object {
  "p1": Object {
    "hp": 7,
  },
  "p2": Object {
    "hp": 7,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 3 power and player 2: ATTACK with 3 power calculates it correctly 2`] = `
Object {
  "p1": Object {
    "hp": 7,
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 3 power and player 2: ATTACK with 3 power calculates it correctly 3`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 7,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 3 power and player 2: ATTACK with 3 power calculates it correctly 4`] = `
Object {
  "p1": Object {
    "hp": 7,
  },
  "p2": Object {
    "hp": 7,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 3 power and player 2: DODGE  calculates it correctly 1`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 3 power and player 2: DODGE  calculates it correctly 2`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 3 power and player 2: PARRY  calculates it correctly 1`] = `
Object {
  "p1": Object {
    "hp": 10,
    "move": Object {
      "turns": 2,
      "type": "STAGGER",
    },
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: ATTACK with 3 power and player 2: PARRY  calculates it correctly 2`] = `
Object {
  "p1": Object {
    "hp": 10,
    "move": Object {
      "turns": 6,
      "type": "STAGGER",
    },
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: DODGE  and player 2: ATTACK with 1 power calculates it correctly 1`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: DODGE  and player 2: ATTACK with 1 power calculates it correctly 2`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: DODGE  and player 2: ATTACK with 1 power calculates it correctly 3`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: DODGE  and player 2: ATTACK with 2 power calculates it correctly 1`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: DODGE  and player 2: ATTACK with 3 power calculates it correctly 1`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: DODGE  and player 2: ATTACK with 3 power calculates it correctly 2`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: DODGE  and player 2: DODGE  calculates it correctly 1`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: DODGE  and player 2: PARRY  calculates it correctly 1`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 10,
    "move": Object {
      "turns": 1,
      "type": "STAGGER",
    },
  },
}
`;

exports[`Battle system when player 1: PARRY  and player 2: ATTACK with 1 power calculates it correctly 1`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: PARRY  and player 2: ATTACK with 1 power calculates it correctly 2`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 10,
    "move": Object {
      "turns": 1,
      "type": "STAGGER",
    },
  },
}
`;

exports[`Battle system when player 1: PARRY  and player 2: ATTACK with 1 power calculates it correctly 3`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 10,
    "move": Object {
      "turns": 1,
      "type": "STAGGER",
    },
  },
}
`;

exports[`Battle system when player 1: PARRY  and player 2: ATTACK with 2 power calculates it correctly 1`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 10,
    "move": Object {
      "turns": 1,
      "type": "STAGGER",
    },
  },
}
`;

exports[`Battle system when player 1: PARRY  and player 2: ATTACK with 3 power calculates it correctly 1`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 10,
    "move": Object {
      "turns": 2,
      "type": "STAGGER",
    },
  },
}
`;

exports[`Battle system when player 1: PARRY  and player 2: ATTACK with 3 power calculates it correctly 2`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 10,
    "move": Object {
      "turns": 6,
      "type": "STAGGER",
    },
  },
}
`;

exports[`Battle system when player 1: PARRY  and player 2: DODGE  calculates it correctly 1`] = `
Object {
  "p1": Object {
    "hp": 10,
    "move": Object {
      "turns": 1,
      "type": "STAGGER",
    },
  },
  "p2": Object {
    "hp": 10,
  },
}
`;

exports[`Battle system when player 1: PARRY  and player 2: PARRY  calculates it correctly 1`] = `
Object {
  "p1": Object {
    "hp": 10,
  },
  "p2": Object {
    "hp": 10,
  },
}
`;
